frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1182 @@
+		) -> DispatchResult {
+			let owner = ensure_signed(origin)?;
+			let admin = T::Lookup::lookup(admin)?;
+
+			ensure!(!Class::<T, I>::contains_key(class), Error::<T, I>::InUse);
+
+			let deposit = T::ClassDeposit::get();
+			T::Currency::reserve(&owner, deposit)?;
+
+			Class::<T, I>::insert(
+				class,
+				ClassDetails {
+					owner: owner.clone(),
+					issuer: admin.clone(),
+					admin: admin.clone(),
+					freezer: admin.clone(),
+					deposit,
+					metadata_deposit: Zero::zero(),
+					free_holding: false,
+					instances: 0,
+					free_holds: 0,
+					is_frozen: false,
+				},
+			);
+			Self::deposit_event(Event::Created(class, owner, admin));
+			Ok(())
+		}
+
+		/// Issue a new class of non-fungible assets from a privileged origin.
+		///
+		/// This new asset class has no assets initially.
+		///
+		/// The origin must conform to `ForceOrigin`.
+		///
+		/// Unlike `create`, no funds are reserved.
+		///
+		/// - `class`: The identifier of the new asset. This must not be currently in use.
+		/// - `owner`: The owner of this class of assets. The owner has full superuser permissions
+		/// over this asset, but may later change and configure the permissions using
+		/// `transfer_ownership` and `set_team`.
+		///
+		/// Emits `ForceCreated` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::force_create())]
+		pub(super) fn force_create(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			owner: <T::Lookup as StaticLookup>::Source,
+			free_holding: bool,
+		) -> DispatchResult {
+			T::ForceOrigin::ensure_origin(origin)?;
+			let owner = T::Lookup::lookup(owner)?;
+
+			ensure!(!Class::<T, I>::contains_key(class), Error::<T, I>::InUse);
+
+			Class::<T, I>::insert(
+				class,
+				ClassDetails {
+					owner: owner.clone(),
+					issuer: owner.clone(),
+					admin: owner.clone(),
+					freezer: owner.clone(),
+					deposit: Zero::zero(),
+					metadata_deposit: Zero::zero(),
+					free_holding,
+					instances: 0,
+					free_holds: 0,
+					is_frozen: false,
+				},
+			);
+			Self::deposit_event(Event::ForceCreated(class, owner));
+			Ok(())
+		}
+
+		/// Destroy a class of fungible assets.
+		///
+		/// The origin must conform to `ForceOrigin` or must be `Signed` and the sender must be the
+		/// owner of the asset `class`.
+		///
+		/// - `class`: The identifier of the asset class to be destroyed.
+		/// - `witness`: Information on the instances minted in the asset class. This must be
+		/// correct.
+		///
+		/// Emits `Destroyed` event when successful.
+		///
+		/// Weight: `O(i + f)` where:
+		/// - `i = (witness.instances - witness.free_holds)`
+		/// - `f = witness.free_holds`
+		#[pallet::weight(T::WeightInfo::destroy(
+			witness.instances.saturating_sub(witness.free_holds),
+ 			witness.free_holds,
+ 		))]
+		pub(super) fn destroy(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			witness: DestroyWitness,
+		) -> DispatchResult {
```

Comment:
Here is an alternative way I would write a destroy function which cleaned up the state of the chain a bit better, and allows for anyone to help.

* A function that sets `asset_pending_destruction: bool`
* If that is set, then anyone can call `try_destroy(class)`
* We loop as far as we can destroying assets until we hit the weight limit
* Then we stop and wait for the next call to `try_destroy`
* If all assets are destroyed, we complete the operation returning any deposits to the right people

This would also allow us to remove metadata along the way rather than having a separate call needed to remove metadata for each unique that has some.

Reply:
This also means we never need to worry about the total volume of the asset being beyond what can be cleaned up, for example if the witness data would push beyond the weight limit of a single block.

Reply:
I'd be ok with a PR which added that functionality.

FWIW:

No separate call is needed for destroying metadata. It already gets cleared along the way.

>for example if the witness data would push beyond the weight limit of a single block

In this case you'd just burn enough the assets manually to get down to the block limit.